Use `details` and `summary` elements in post metaboxes and dashboard widgets

KTS915 opened the pull request:
This PR replaces the current post metaboxes and dashboard widgets (which are rendered as `divs` and toggled via JavaScript) with `details and `summary` elements which enhance accessibility and require no JavaScript for toggling.

## Motivation and context
Short-term: to enhance accessibility, reduce reliance on JavaScript, and bring the code more into line with modern HTML5
Longer term: to facilitate eventual replacement of outdated jQuery-based code and libraries like jQuery UI

## How has this been tested?
On my localhost install

## Screenshots
These aren't quite like-for-like, as they're from different sites, but they should make things clear.

### Before
![Screenshot at 2023-09-19 11-56-51](https://github.com/ClassicPress/ClassicPress-v2/assets/4127662/1523044a-8a0c-4548-8f98-b94246e424cc)
![Screenshot at 2023-09-19 11-57-17](https://github.com/ClassicPress/ClassicPress-v2/assets/4127662/99098737-6a69-4450-b877-1eff7669644a)

### After
![Screenshot at 2023-09-19 11-59-46](https://github.com/ClassicPress/ClassicPress-v2/assets/4127662/c9fa5fa4-421a-4702-99cb-f7e9c9179ef3)
![Screenshot at 2023-09-19 12-00-08](https://github.com/ClassicPress/ClassicPress-v2/assets/4127662/23cd5e88-7b6a-44cd-a742-b88b70dd8e83)

## Types of changes
New feature: not a breaking change because it respects the pre-existing settings of open or closed.

KTS915 committed:
Adds details and summary elements to post metaboxes and dashboard widgets

This enhances accessibility for post metaboxes and dashboard widgets and enables them to be toggled open and closed without JavaScript.

KTS915 committed:
Adds details and summary elements to post metaboxes and dashboard widgets

This enhances accessibility for post metaboxes and dashboard widgets and enables them to be toggled open and closed without JavaScript.

KTS915 committed:
Update edit.css

Ensures that post metaboxes and dashboard widgets look the same as before the use of details and summary elements.

KTS915 committed:
Update edit.css

Ensures that post metaboxes and dashboard widgets look the same as before the use of details and summary elements.

KTS915 committed:
Update common.css

Enures that post metaboxes and dashboard widgets look the same as before the use of details and summary elements.

KTS915 committed:
Update common.css

Enures that post metaboxes and dashboard widgets look the same as before the use of details and summary elements.

KTS915 committed:
Update postbox.js

Ensures that open/closed state of post metaboxes and dashboard widgets is preserved across page loads. Before it was based on the presence or otherwise of the class "closed"; now it is dependent on the presence or otherwise of the attribute "open".

KTS915 committed:
Update postbox.js

Ensures that open/closed state of post metaboxes and dashboard widgets is preserved across page loads. Before it was based on the presence or otherwise of the class "closed"; now it is dependent on the presence or otherwise of the attribute "open".

KTS915 committed:
Update sortable.js

Sortable widgets in the CP backend currently rely on jQuery UI. While it would be nice to replace that with something more modern (like SortableJS), that has to wait for another day. In the meantime, we need a fix to address the fact that the HTML5 elements `details` and `summary` were introduced after development ceased on jQuery UI.

This fix simply avoids the presence of an ugly artefact when dragging and dropping a `details` element.

KTS915 committed:
Update sortable.js

Sortable widgets in the CP backend currently rely on jQuery UI. While it would be nice to replace that with something more modern (like SortableJS), that has to wait for another day. In the meantime, we need a fix to address the fact that the HTML5 elements `details` and `summary` were introduced after development ceased on jQuery UI.

This fix simply avoids the presence of an ugly artefact when dragging and dropping a `details` element.

mattyrob committed:
WP-r56622: Add `sys_get_temp_dir()` to `open_basedir` tests. (#218)

In PHPUnit 10.3.5, 9.6.13 and 8.5.34, the child processes used for process isolation now use temporary files to communicate their result to the parent process.

This caused a failure in some tests that set the `open_basedir` PHP directive to a value that did not include `sys_get_temp_dir()`.

This adds `sys_get_temp_dir()` to the `open_basedir` value set by the tests to ensure that permission is still granted for the temporary directory.

PHPUnit uses `sys_get_temp_dir()`. To ensure the result is the same, Core's `get_temp_dir()` function is not used.

References:
- https://github.com/sebastianbergmann/phpunit/issues/5356

WP:Props desrosj, mukesh27, SergeyBiryukov, costdev.
Fixes https://core.trac.wordpress.org/ticket/59394.

---

Merges https://core.trac.wordpress.org/changeset/56622 / WordPress/wordpress-develop@7d96189ba1 to ClassicPress.

Co-authored-by: Colin Stewart <costdev@git.wordpress.org>

renovate[bot] committed:
Renovate[bot]: Update dependency node to v18.18.0 (#216)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

mattyrob committed:
Merge branch 'meta-boxes-dashboard-widgets' of https://github.com/KTS915/ClassicPress-v2 into pr/217

mattyrob committed:
PHP Coding standards

mattyrob committed:
JS Coding Standards

rkingett commented:
This is a really great change! This will work in Braille too which will also be good!

xxsimoxx reviewed (commented):
When over the handle the cursor is `move` and if clicked it collapse the item, if dragged it moves the item.
My idea is that it should not be draggable and have the cursor of type `pointer`.

extra808 commented:
`Details/summary` has significant accessibility bugs in Safari.

In Safari for Mac (versions 16 and 17), [all focusable elements except links remain focusable](https://bugs.webkit.org/show_bug.cgi?id=260523) when the `details` is collapsed. This is not a VoiceOver issue, they're correctly hidden to it, it's a sighted keyboard navigation issue.

In mobile Safari (iOS, iPadOS version 15, 16, and 17) their [role and state are not announced by VoiceOver.](https://bugs.webkit.org/show_bug.cgi?id=252223#c4) If you correctly guess that the `summary` is an operable control, it does open/close the `details` but it's only apparent if you can see it.

KTS915 commented:
@xxsimoxx Thanks, I will do that.

@extra808 Thanks for these comments! I was under the impression that users who wanted accessibility enhancements these days use either Edge or Firefox as their browser.

I can certainly add some JavaScript to address the focusable issue. As for the latter issue, the discussion to which you've linked suggests this is a relatively recent regression in Safari. Do you know what the plans are for addressing it?

KTS915 committed:
Update common.js to address Safari's failure to respect effect of closure of details element

Addresses the focusable issue mentioned in this comment: https://github.com/ClassicPress/ClassicPress-v2/pull/217#issuecomment-1735439355

This commit applies the `inert` attribute to elements within a closed `details` element. This appears to be supported by both desktop and mobile versions of Safari: see https://caniuse.com/?search=inert

KTS915 commented:
I have just submitted a patch that addresses the focusable issue that @extra808 mentioned. I have added the relevant code to the `common.js` file so that it will work with all `details` elements throughout the admin.

KTS915 committed:
Update common.css to replace details arrows via CSS so that they can also have a pointer cursor.

KTS915 commented:
@xxsimoxx I've just pushed a patch that replaces the default arrows with CSS-generated arrows. Then I have applied `cursor: pointer` to those arrows. The end result is to replicate exactly how the metaboxes worked before: hovering over the title of the metabox gets a draggable icon for the cursor, whereas hovering over the arrow gets a pointer icon for the cursor.

KTS915 commented:
@extra808 I have another question for you. In mobile Safari with things as they are (i.e. without the changes proposed here), can the user drag the metaboxes?

KTS915 committed:
Update common.css to improve style of arrows

xxsimoxx reviewed (approved):
Looks good!

mattyrob committed:
JS syntax corrections

extra808 commented:
> @extra808 Thanks for these comments! I was under the impression that users who wanted accessibility enhancements these days use either Edge or Firefox as their browser.

I don't know what you mean by "accessibility enhancements." People who use VoiceOver for Mac almost exclusively use Safari because that is the most compatible combination; for some specific content, other browsers may do better, for example VoiceOver+Firefox does not have either of these details/summary problems. Of course on iOS/iPadOS, users have no choice but to use Safari/WebKit.

> I can certainly add some JavaScript to address the focusable issue. As for the latter issue, the discussion to which you've linked suggests this is a relatively recent regression in Safari. Do you know what the plans are for addressing it?

Both issues are regressions, the focus one is the more recent regression. I linked to WebKit's public bug tracker, which is not secretive like Apple's internal one, but it's still common for bugs to go unacknowledged and unaddressed for years. But I don't think reports in Firefox's or Chrome's bug trackers are handled any better.

extra808 commented:
> @extra808 I have another question for you. In mobile Safari with things as they are (i.e. without the changes proposed here), can the user drag the metaboxes?

I don't know, I don't use ClassicPress, I just saw a post on Mastodon about this change and thought I'd share what I knew about issues with the elements. 
I really like details/summary and I'm frustrated by browser implementation problems.

mattyrob commented:
Considered for merge today but on testing I noticed that the toggle arrow is quite small now on FF and Chrome.
<img width="282" alt="Screenshot 2023-09-28 at 16 28 11" src="https://github.com/ClassicPress/ClassicPress-v2/assets/1280733/ac96b5e2-a62d-419d-a0c6-a717db99f300">

And in Safari 17 the arrow is duplicated:
<img width="286" alt="Screenshot 2023-09-28 at 16 28 26" src="https://github.com/ClassicPress/ClassicPress-v2/assets/1280733/29d56882-2c11-4785-b1ba-ac4f5c46b2a6">

This patterns repeats through all areas of implementation.

KTS915 committed:
Update common.css to make `details` marker bigger and remove the default in Safari.

KTS915 commented:
> Considered for merge today but on testing I noticed that the toggle arrow is quite small now on FF and Chrome. <img alt="Screenshot 2023-09-28 at 16 28 11" width="282" src="https://user-images.githubusercontent.com/1280733/271336880-ac96b5e2-a62d-419d-a0c6-a717db99f300.png">
> 
> And in Safari 17 the arrow is duplicated: <img alt="Screenshot 2023-09-28 at 16 28 26" width="286" src="https://user-images.githubusercontent.com/1280733/271336957-29d56882-2c11-4785-b1ba-ac4f5c46b2a6.png">
> 
> This patterns repeats through all areas of implementation.

Both of these issues should now be addressed by the latest commit.

mattyrob committed:
Compliant CSS comment

mattyrob commented:
Arrow size is much better, but still seeing two in Safari:

<img width="284" alt="Screenshot 2023-09-29 at 18 05 03" src="https://github.com/ClassicPress/ClassicPress-v2/assets/1280733/b5dd2005-db5e-42c8-9af1-72b34d7e737a">

<img width="293" alt="Screenshot 2023-09-29 at 18 05 14" src="https://github.com/ClassicPress/ClassicPress-v2/assets/1280733/5a11e9a0-2bdf-48bd-a56b-eb8e1f872f8d">

extra808 commented:
It looks like Safari needs a separate rule, it ignores `list-style-type: none` for `summary`. The [MDN page on `summary`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/summary#default_style) mentions this.

```CSS
/* Replace default marker in details element */
.metabox-holder details > summary {
  list-style-type: none;
}
.metabox-holder details > summary::-webkit-details-marker {
  display: none;
}
```

mattyrob committed:
Suggested Safari fix

rkingett commented on src/wp-admin/css/common.css:
```diff
@@ -2061,14 +2061,13 @@ html.wp-toolbar {
 }
 
 /* Replace default marker in details element */
-.metabox-holder details > summary,
```
This is having additional unforseen consequences. Mainly, the navigational headings that served as landmarks are now removed from the boxes such as post, categories, and the like. I tried to pinpoint where this was happening but wasn't quite sure where to leave this.

KTS915 replied:
@rkingett Is this something that only happened with the change you've linked to, or did it happen before? I'm asking because I'm wondering whether it's actually been caused by my removing the `h2` and `h3` tags in my original commit. (I did that after reading conflicting advice on whether I should or not.)

I just want to be sure where the problem lies so that we know where to fix it.

KTS915 replied:
I've just pushed a couple of commits to add back the `h2 and `h3` tags.

KTS915 committed:
Update common.css for summary header and custom marker

KTS915 committed:
Update template.php to add back `h2` and `h3` tags.

This responds to feedback at https://github.com/ClassicPress/ClassicPress-v2/pull/217#discussion_r1341996477

This should also resolve the issue where clicking on the pointer icon still enabled the widget to be dragged. Dragging is now possible only where the draggable icon appears.

KTS915 committed:
Update template.php to move a class to avoid causing problems on the admin menus page.

KTS915 committed:
Update common.css to modify styling on admin menus page.

The changes made elsewhere in the PR have an impact on the visual styling of the admin menus page. For reasons long forgotten, and as noted in this file already, that page works differently from other admin pages that use widgets. So some modifications are necessary to ensure the menus page continues to look as expected.

KTS915 committed:
Update template.php to remove tabindex and enqueuing of deleted script.

mattyrob committed:
Fix whitespace in wp-admin/includes/template.php
